runner/src/main/kotlin/com/cosmotech/runner/service/RunnerService.kt

```diff
@@ -256,6 +287,57 @@ class RunnerService(
               ?: mutableListOf())
     }
 
+    @Suppress("NestedBlockDepth")
+    private fun constructParametersValuesFromRunTemplateAndParent(
+        parentId: String?,
+        solution: Solution?,
+        runTemplate: RunTemplate?,
+        parent: Runner,
+        runner: Runner
+    ): MutableList<RunnerRunTemplateParameterValue> {
+      val parametersValuesList = mutableListOf<RunnerRunTemplateParameterValue>()
+      logger.debug("Copying parameters values from parent $parentId")
+
+      logger.debug("Getting runTemplate parameters ids")
+      val runTemplateParametersIds =
+          solution
+              ?.parameterGroups
+              ?.filter { parameterGroup ->
+                runTemplate?.parameterGroups?.contains(parameterGroup.id) == true
+              }
+              ?.flatMap { parameterGroup -> parameterGroup.parameters ?: mutableListOf() }
+      if (!runTemplateParametersIds.isNullOrEmpty()) {
+        val parentParameters = parent.parametersValues?.associate { it.parameterId to it }
+        val runnerParameters = runner.parametersValues?.associate { it.parameterId to it }
+        // TODO Handle default value
+        runTemplateParametersIds.forEach { parameterId ->
+          if (runnerParameters?.contains(parameterId) != true) {
+            logger.debug(
+                "Parameter $parameterId is not defined in the Runner. " +
+                    "Checking if it is defined in its parent $parentId")
+            if (parentParameters?.contains(parameterId) == true) {
+              logger.debug("Copying parameter value from parent for parameter $parameterId")
+              val parameterValue = parentParameters[parameterId]
+              if (parameterValue != null) {
+                parameterValue.isInherited = true
+                parametersValuesList.add(parameterValue)
+              } else {
+                logger.warn(
+                    "Parameter $parameterId not found in parent ($parentId) parameters values")
+              }
+            } else {
+              logger.debug(
+                  "Skipping parameter $parameterId, defined neither in the parent nor in this Runner")
+            }
+          } else {
+            logger.debug(
+                "Skipping parameter $parameterId since it is already defined in this Runner")
+          }
```

Comment:
I'm not a big fan on nested if-else for error handling. I find sequential if-return easier to read and understand.

Reply:
Rewrite using filters lambda ;)